pkg/logs/input/container/scanner.go

```diff
@@ -3,207 +3,30 @@
-	for _, source := range sources {
-		switch source.Config.Type {
-		case config.DockerType:
-			containerSources = append(containerSources, source)
-		default:
+// NewScanner returns a new container scanner.
+func NewScanner(sources *config.LogSources, pp pipeline.Provider, auditor *auditor.Auditor) restart.Restartable {
+	if config.LogsAgent.GetBool("logs_config.container_collect_all") {
+		if scanner, err := NewKubeScanner(sources, pp, auditor); err == nil {
+			// Fow now, avoid manually scanning docker containers when in a
+			// kubernetes environment, and rely on Kubernetes API.
+			return scanner
 		}
-	}
-	return &Scanner{
-		pp:      pp,
-		sources: containerSources,
-		tailers: make(map[string]*DockerTailer),
-		auditor: a,
-		stop:    make(chan struct{}),
-	}
-}
-
-// Start starts the Scanner
-func (s *Scanner) Start() {
-	err := s.setup()
-	if err != nil {
-		s.reportErrorToAllSources(err)
-		return
-	}
-	go s.run()
-	s.isRunning = true
-}
-
-// Stop stops the Scanner and its tailers in parallel,
-// this call returns only when all the tailers are stopped
-func (s *Scanner) Stop() {
-	if !s.isRunning {
-		// the scanner was not start, no need to stop anything
-		return
-	}
-	s.stop <- struct{}{}
-	stopper := restart.NewParallelStopper()
-	for _, tailer := range s.tailers {
-		stopper.Add(tailer)
-		delete(s.tailers, tailer.ContainerID)
-	}
-	stopper.Stop()
-}
-
-// run checks periodically which docker containers are running until stop
-func (s *Scanner) run() {
-	scanTicker := time.NewTicker(scanPeriod)
-	defer scanTicker.Stop()
-	for {
-		select {
-		case <-scanTicker.C:
-			// check all the containers running on the host and start new tailers if needed
-			s.scan(true)
-		case <-s.stop:
-			// no docker container should be tailed anymore
-			return
-		}
-	}
-}
-
-// scan checks for new containers we're expected to
-// tail, as well as stopped containers or containers that
-// restarted
-func (s *Scanner) scan(tailFromBeginning bool) {
-	runningContainers := s.listContainers()
-	containersToMonitor := make(map[string]bool)
-
-	// monitor new containers, and restart tailers if needed
-	for _, container := range runningContainers {
-		source := NewContainer(container).findSource(s.sources)
-		if source == nil {
-			continue
+		// Append a fake source to collect all logs from all containers.
+		dockerConfig := &config.LogsConfig{
+			Type:    config.DockerType,
+			Service: "docker",
+			Source:  "docker",
 		}
-		tailer, isTailed := s.tailers[container.ID]
-		if isTailed && tailer.shouldStop {
-			continue
-		}
-		if !isTailed {
-			// setup a new tailer
-			succeeded := s.setupTailer(s.cli, container, source, tailFromBeginning, s.pp.NextPipelineChan())
-			if !succeeded {
-				// the setup failed, let's try to tail this container in the next scan
-				continue
-			}
-		}
-		containersToMonitor[container.ID] = true
+		sources.AddSource(config.NewLogSource("container_collect_all", dockerConfig))
```

Comment:
we're altering the original list of sources, could that be a problem in the future?